### PR TITLE
Add Devin CLI harness tracing to Arize AX / Phoenix

### PR DESCRIPTION
**Pull request by duncankmckinnon**
## Summary

Adds a new self-contained `tracing/devin/` harness that traces Devin CLI
sessions to Arize AX / Phoenix as OpenInference spans. Because Devin hook
payloads are thin and token metrics only exist once a session finishes, the
harness registers a single `SessionEnd` command hook that parses the local
ATIF-v1.7 transcript and emits the full span tree (session root, per-step LLM
spans with real token counts, and tool spans) in one deferred pass.

## Changes

- Add `tracing/devin/` package: `constants.py`, `transcript.py` (ATIF-v1.7
  parser), `hooks/adapter.py` (session/transcript resolution via `sessions.db`
  and `DEVIN_PROJECT_DIR`, idempotency), and `hooks/handlers.py` (SessionEnd
  handler building the OpenInference span tree).
- Add `tracing/devin/install.py` and `core/setup/devin.py` to register the
  `SessionEnd` hook and route setup through the shared installer helpers.
- Wire the harness into packaging and installers: `pyproject.toml` console
  script, `install.sh`, and `install.bat`.
- Add `tracing/devin/README.md` and a `README.md` entry documenting the
  integration.
- Add tests under `tests/tracing/devin/` covering the transcript parser (including
  edge cases), adapter resolution/polling, the SessionEnd handler span emission,
  and installer entry points, plus an ATIF-v1.7 fixture.

## Test plan

- [x] `uv run pytest tests/tracing/devin/` passes.
- [x] `uv run pre-commit run --all-files` is clean.
- [x] Installer registers the `SessionEnd` hook in `~/.config/devin/config.json`
      and the `arize-setup-devin` console script is available after install.
- [x] Parser reads a real ATIF-v1.7 transcript and produces LLM spans with
      prompt/completion/total token counts and cache_read on prompt_details.
- [x] A completed Devin session emits a session root span with per-step LLM and
      tool spans to the configured backend; a foreign SessionEnd payload is a no-op.
- [x] Uninstall removes the harness hook entry cleanly.

🤖 Generated by workbench pr_writer

### CODE REVIEW

File: tracing/devin/install.py [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+def _unregister_hooks() -> None:
+    """Remove only our command from each event in ``HOOK_EVENT_NAMES``.
+
+    Drops an event list if it becomes empty, and drops the ``hooks`` key if it
+    becomes empty. Everything else in the config is preserved.
+    """
+    if not CONFIG_FILE.exists():
+        return
+    try:
+        config = json.loads(CONFIG_FILE.read_text())
```
Comment (sanchitram1) [inline tracing/devin/install.py]: devin supports [commented JSON](https://docs.devin.ai/cli/extensibility/configuration) in config files, but json.loads will return `{}` on commented JSON => any config prior to installing coding harness tracing will be lost

File: tracing/devin/constants.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+SERVICE_NAME = "devin"
+SCOPE_NAME = "arize-devin-plugin"
+DEFAULT_PROJECT_NAME = "devin"
+
+# Soft install detection.
+HARNESS_HOME = ".config/devin"  # ~/.config/devin — presence check
+HARNESS_BIN = "devin"  # binary name for shutil.which() fallback
+
+# Global hook config file (hooks nest under a top-level "hooks" key here).
+CONFIG_FILE = Path.home() / ".config" / "devin" / "config.json"
```
Comment (sanchitram1) [inline tracing/devin/constants.py]: won't this be different for Windows? from [their docs](https://docs.devin.ai/cli/reference/configuration/config-file):

> On Windows, the user config path is %APPDATA%\devin\config.json (e.g. C:\Users\<you>\AppData\Roaming\devin\config.json), not ~\.config\devin\config.json.

File: tracing/devin/hooks/handlers.py [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+            SCOPE_NAME,
+        )
+        send_span(step_span)
+
+        for tc in step.tool_calls:
+            tool_attrs: dict[str, Any] = {
+                "session.id": session_id,
+                "openinference.span.kind": "TOOL",
+                "tool.name": tc.name,
+                "input.value": redact_content(env.log_tool_content, json.dumps(tc.arguments)),
```
Comment (sanchitram1) [inline tracing/devin/hooks/handlers.py]: I think this is supposed to be `env.log_tool_details`, not tool content. content protects the outputs, while details protects the inputs – is my understanding

**Review by sanchitram1 — CHANGES_REQUESTED** [review]
- windows config path
- json.loads drops commented JSON (which devin supports) 
- swap log_tool_content for log_tools_detail in handlers.py:177

File: tests/tracing/devin/fixtures/session.json [hunk 4]
```diff
@@ -0,0 +1,25 @@
+{
```
Comment (sanchitram1) [inline tests/tracing/devin/fixtures/session.json]: nit: dead fixture, can delete it

File: tracing/devin/session_db.py [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+
+
+def connect_readonly(db_path, retries: int = 3, delay: float = 0.1):
+    """Open ``db_path`` read-only, honoring the WAL (so the newest, not-yet-
+    checkpointed rows are visible).
+
+    We deliberately do NOT use ``immutable=1``: it tells SQLite the file cannot
+    change and to ignore the ``-wal`` sidecar, which would hide the very rows a
+    just-fired ``Stop`` cares about. ``mode=ro`` respects the WAL. Retries a few
+    times on transient lock errors. Returns ``None`` on failure.
```
Comment (sanchitram1) [inline tracing/devin/session_db.py]: nit: this raises an exception on failure

File: tracing/devin/hooks/adapter.py [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    than a silently dropped trace).
+    """
+    if not session_id or not request_id:
+        return False
+    key = _emit_key(session_id, request_id)
+    path = _emitted_file()
+    try:
+        if not path.exists():
+            return False
+        with path.open("r", encoding="utf-8") as fh:
```
Comment (sanchitram1) [inline tracing/devin/hooks/adapter.py:69]: if we perceive degraded performance with `devin`, it would probably be here, since this scan is O(n^2) since it happens per step per flush. in the future, we can consider pruning entries, or loading the file once per flush into a set.

File: tests/tracing/devin/test_devin_session_db.py [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+# ---------------------------------------------------------------------------
+# resolve_session_id / connect_readonly
+# ---------------------------------------------------------------------------
+
+
+def test_resolve_session_id_returns_newest(tmp_path):
+    db = tmp_path / "sessions.db"
+    _build_db(
+        db,
+        [("old", "/work/proj", "Windsurf", "m", 100, 1), ("new", "/work/proj", "Windsurf", "m", 200, 1)],
```
Comment (sanchitram1) [inline tests/tracing/devin/test_devin_session_db.py:102]: i'd add cases with trailing slashes here, symlinks, case insensitivity (for Windows, `c:` and `C:` are different)

I think the right approach is to modify the underlying function to use Path.resolve

File: tracing/devin/session_db.py [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+    return None
+
+
+def resolve_session_id(con: sqlite3.Connection, project_dir: str) -> str | None:
+    """Most recent session id whose ``working_directory`` matches ``project_dir``."""
+    if not project_dir:
+        return None
+    try:
+        row = con.execute(
+            "SELECT id FROM sessions WHERE working_directory = ? ORDER BY last_activity_at DESC LIMIT 1",
```
Comment (sanchitram1) [inline tracing/devin/session_db.py]: grab the IDs and use Path.resolve or os.path.realpath to do the comparison, which avoids a lot of common edge cases

**Review by sanchitram1 — CHANGES_REQUESTED** [review]
mostly nits, but we should fix the Path comparison: `session_db.py:resolve_session_id` before merging

**Review by musicionary — APPROVED** [review]

**Review by sanchitram1 — APPROVED** [review]
